### DISCUSSION TITLE
How to run Pandas Tutor?

### DISCUSSION THREAD

**Discussion by jothikannan89** [pyodide/pyodide#3884] [General]
I am trying to run the pandas tutor with pyodide like explained here https://blog.pyodide.org/posts/pandastutor/#the-future-pandas-tutor--pyodide-everywhere, but it says `run_user_code` is not found, and I found this method is not available. Instead of I tried to run `parse_as_json` and its working but it is not retuning the trace code, instead it returning some other statements in the json

Trace code I am getting from parse_as_json

```
{
  "type_": "ParsedModule",
  "code": "import pandas as pd\nimport io\n\ncsv = '''\nbreed,type,longevity,size,weight\nGerman Shepherd,herding,9.73,large,\nBeagle,hound,12.3,small,\nYorkshire Terrier,toy,12.6,small,5.5\nGolden Retriever,sporting,12.04,medium,60.0\nBulldog,non-sporting,6.29,medium,45.0\nLabrador Retriever,sporting,12.04,medium,67.5\nBoxer,working,8.81,medium,\nPoodle,non-sporting,11.95,medium,\nDachshund,hound,12.63,small,24.0\nRottweiler,working,9.11,large,\nBoston Terrier,non-sporting,10.92,medium,\nShih Tzu,toy,13.2,small,12.5\nMiniature Schnauzer,terrier,11.81,small,15.5\nDoberman Pinscher,working,10.33,large,\nChihuahua,toy,16.5,small,5.5\nSiberian Husky,working,12.58,medium,47.5\nPomeranian,toy,9.67,small,5.0\nFrench Bulldog,non-sporting,9.0,medium,27.0\nGreat Dane,working,6.96,large,\nShetland Sheepdog,herding,12.53,small,22.0\nCavalier King Charles Spaniel,toy,11.29,small,15.5\nGerman Shorthaired Pointer,sporting,11.46,large,62.5\nMaltese,toy,12.25,small,5.0\n'''\n\ndogs = pd.read_csv(io.StringIO(csv))\n\n(dogs[dogs['size'] == 'medium']\n    .sort_values('type')\n    .groupby('type').median()\n)\n",
  "location": {
    "start": {
      "line": 0,
      "ch": 0
    },
    "end": {
      "line": 36,
      "ch": 0
    }
  },
  "statements": [
    {
      "type_": "VerbatimStatement",
      "code": "import pandas as pd",
      "location": {
        "start": {
          "line": 0,
          "ch": 0
        },
        "end": {
          "line": 0,
          "ch": 19
        }
      }
    },
    {
      "type_": "VerbatimStatement",
      "code": "import io",
      "location": {
        "start": {
          "line": 1,
          "ch": 0
        },
        "end": {
          "line": 1,
          "ch": 9
        }
      }
    },
    {
      "type_": "ChainStatement",
      "code": "csv = '''\nbreed,type,longevity,size,weight\nGerman Shepherd,herding,9.73,large,\nBeagle,hound,12.3,small,\nYorkshire Terrier,toy,12.6,small,5.5\nGolden Retriever,sporting,12.04,medium,60.0\nBulldog,non-sporting,6.29,medium,45.0\nLabrador Retriever,sporting,12.04,medium,67.5\nBoxer,working,8.81,medium,\nPoodle,non-sporting,11.95,medium,\nDachshund,hound,12.63,small,24.0\nRottweiler,working,9.11,large,\nBoston Terrier,non-sporting,10.92,medium,\nShih Tzu,toy,13.2,small,12.5\nMiniature Schnauzer,terrier,11.81,small,15.5\nDoberman Pinscher,working,10.33,large,\nChihuahua,toy,16.5,small,5.5\nSiberian Husky,working,12.58,medium,47.5\nPomeranian,toy,9.67,small,5.0\nFrench Bulldog,non-sporting,9.0,medium,27.0\nGreat Dane,working,6.96,large,\nShetland Sheepdog,herding,12.53,small,22.0\nCavalier King Charles Spaniel,toy,11.29,small,15.5\nGerman Shorthaired Pointer,sporting,11.46,large,62.5\nMaltese,toy,12.25,small,5.0\n'''",
      "location": {
        "start": {
          "line": 3,
          "ch": 0
        },
        "end": {
          "line": 28,
          "ch": 3
        }
      },
      "chain": []
    },
    {
      "type_": "ChainStatement",
      "code": "dogs = pd.read_csv(io.StringIO(csv))",
      "location": {
        "start": {
          "line": 30,
          "ch": 0
        },
        "end": {
          "line": 30,
          "ch": 36
        }
      },
      "chain": [
        {
          "type_": "StartOfChain",
          "code": "pd",
          "location": {
            "start": {
              "line": 30,
              "ch": 7
            },
            "end": {
              "line": 30,
              "ch": 9
            }
          }
        },
        {
          "type_": "PassThroughCall",
          "code": "pd.read_csv(io.StringIO(csv))",
          "location": {
            "start": {
              "line": 30,
              "ch": 9
            },
            "end": {
              "line": 30,
              "ch": 36
            }
          },
          "func": "read_csv"
        }
      ]
    },
    {
      "type_": "ChainStatement",
      "code": "(dogs[dogs['size'] == 'medium']\n    .sort_values('type')\n    .groupby('type').median()\n)",
      "location": {
        "start": {
          "line": 32,
          "ch": 0
        },
        "end": {
          "line": 35,
          "ch": 1
        }
      },
      "chain": [
        {
          "type_": "StartOfChain",
          "code": "dogs",
          "location": {
            "start": {
              "line": 32,
              "ch": 1
            },
            "end": {
              "line": 32,
              "ch": 5
            }
          }
        },
        {
          "type_": "Subscript",
          "code": "dogs[dogs['size'] == 'medium']",
          "location": {
            "start": {
              "line": 32,
              "ch": 5
            },
            "end": {
              "line": 32,
              "ch": 31
            }
          },
          "slicer": null,
          "slice1": {
            "type_": "SubsComparison",
            "code": "dogs['size'] == 'medium'",
            "location": {
              "start": {
                "line": 32,
                "ch": 6
              },
              "end": {
                "line": 32,
                "ch": 30
              }
            },
            "label_exprs": [
              "'size'"
            ]
          },
          "slice2": null
        },
        {
          "type_": "SortValuesCall",
          "code": "dogs[dogs['size'] == 'medium']\n    .sort_values('type')",
          "location": {
            "start": {
              "line": 33,
              "ch": 4
            },
            "end": {
              "line": 33,
              "ch": 24
            }
          },
          "label_expr": "'type'",
          "axis": "index"
        },
        {
          "type_": "GroupByCall",
          "code": "dogs[dogs['size'] == 'medium']\n    .sort_values('type')\n    .groupby('type')",
          "location": {
            "start": {
              "line": 34,
              "ch": 4
            },
            "end": {
              "line": 34,
              "ch": 20
            }
          },
          "axis": "index"
        },
        {
          "type_": "AggCall",
          "code": "(dogs[dogs['size'] == 'medium']\n    .sort_values('type')\n    .groupby('type').median()\n)",
          "location": {
            "start": {
              "line": 34,
              "ch": 20
            },
            "end": {
              "line": 34,
              "ch": 29
            }
          }
        }
      ]
    }
  ]
}
```

Actual trace code from https://pandastutor.com/

`{"code":"import pandas as pd\nimport io\n\ncsv = '''\nbreed,type,longevity,size,weight\nGerman Shepherd,herding,9.73,large,\nBeagle,hound,12.3,small,\nYorkshire Terrier,toy,12.6,small,5.5\nGolden Retriever,sporting,12.04,medium,60.0\nBulldog,non-sporting,6.29,medium,45.0\nLabrador Retriever,sporting,12.04,medium,67.5\nBoxer,working,8.81,medium,\nPoodle,non-sporting,11.95,medium,\nDachshund,hound,12.63,small,24.0\nRottweiler,working,9.11,large,\nBoston Terrier,non-sporting,10.92,medium,\nShih Tzu,toy,13.2,small,12.5\nMiniature Schnauzer,terrier,11.81,small,15.5\nDoberman Pinscher,working,10.33,large,\nChihuahua,toy,16.5,small,5.5\nSiberian Husky,working,12.58,medium,47.5\nPomeranian,toy,9.67,small,5.0\nFrench Bulldog,non-sporting,9.0,medium,27.0\nGreat Dane,working,6.96,large,\nShetland Sheepdog,herding,12.53,small,22.0\nCavalier King Charles Spaniel,toy,11.29,small,15.5\nGerman Shorthaired Pointer,sporting,11.46,large,62.5\nMaltese,toy,12.25,small,5.0\n'''\n\ndogs = pd.read_csv(io.StringIO(csv))\n\n(dogs[dogs['size'] == 'medium']\n .sort_values('type')\n .groupby('type').median()\n)","trace":[{"tables":{"lhs":{"data":[["German Shepherd","herding",9.73,"large",null],["Beagle","hound",12.3,"small",null],["Yorkshire Terrier","toy",12.6,"small",5.5],["Golden Retriever","sporting",12.04,"medium",60],["Bulldog","non-sporting",6.29,"medium",45],["Labrador Retriever","sporting",12.04,"medium",67.5],["Boxer","working",8.81,"medium",null],["Poodle","non-sporting",11.95,"medium",null],["Dachshund","hound",12.63,"small",24],["Rottweiler","working",9.11,"large",null],["Boston Terrier","non-sporting",10.92,"medium",null],["Shih Tzu","toy",13.2,"small",12.5],["Miniature Schnauzer","terrier",11.81,"small",15.5],["Doberman Pinscher","working",10.33,"large",null],["Chihuahua","toy",16.5,"small",5.5],["Siberian Husky","working",12.58,"medium",47.5],["Pomeranian","toy",9.67,"small",5],["French Bulldog","non-sporting",9,"medium",27],["Great Dane","working",6.96,"large",null],["Shetland Sheepdog","herding",12.53,"small",22],["Cavalier King Charles Spaniel","toy",11.29,"small",15.5],["German Shorthaired Pointer","sporting",11.46,"large",62.5],["Maltese","toy",12.25,"small",5]],"col_labels":["breed","type","longevity","size","weight"],"row_labels":[0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22]},"rhs":{"data":[["Golden Retriever","sporting",12.04,"medium",60],["Bulldog","non-sporting",6.29,"medium",45],["Labrador Retriever","sporting",12.04,"medium",67.5],["Boxer","working",8.81,"medium",null],["Poodle","non-sporting",11.95,"medium",null],["Boston Terrier","non-sporting",10.92,"medium",null],["Siberian Husky","working",12.58,"medium",47.5],["French Bulldog","non-sporting",9,"medium",27]],"col_labels":["breed","type","longevity","size","weight"],"row_labels":[3,4,5,6,7,10,15,17]}},"expr":{"type":"Subscript","code":"(dogs[dogs['size'] == 'medium']\n .sort_values('type')\n .groupby('type').median()\n)","range":{"start":{"line":0,"ch":5},"end":{"line":0,"ch":31}}},"annotations":[{"type":"highlight","target":{"table":"rhs","row":"all","col":3}},{"type":"highlight","target":{"table":"lhs","row":"all","col":3}},{"type":"arrow","from":{"table":"lhs","row":3,"col":"all"},"to":{"table":"rhs","row":0,"col":"all"}},{"type":"arrow","from":{"table":"lhs","row":4,"col":"all"},"to":{"table":"rhs","row":1,"col":"all"}},{"type":"arrow","from":{"table":"lhs","row":5,"col":"all"},"to":{"table":"rhs","row":2,"col":"all"}},{"type":"arrow","from":{"table":"lhs","row":6,"col":"all"},"to":{"table":"rhs","row":3,"col":"all"}},{"type":"arrow","from":{"table":"lhs","row":7,"col":"all"},"to":{"table":"rhs","row":4,"col":"all"}},{"type":"arrow","from":{"table":"lhs","row":10,"col":"all"},"to":{"table":"rhs","row":5,"col":"all"}},{"type":"arrow","from":{"table":"lhs","row":15,"col":"all"},"to":{"table":"rhs","row":6,"col":"all"}},{"type":"arrow","from":{"table":"lhs","row":17,"col":"all"},"to":{"table":"rhs","row":7,"col":"all"}}]},{"tables":{"lhs":"prev_rhs","rhs":{"data":[["Bulldog","non-sporting",6.29,"medium",45],["Poodle","non-sporting",11.95,"medium",null],["Boston Terrier","non-sporting",10.92,"medium",null],["French Bulldog","non-sporting",9,"medium",27],["Golden Retriever","sporting",12.04,"medium",60],["Labrador Retriever","sporting",12.04,"medium",67.5],["Boxer","working",8.81,"medium",null],["Siberian Husky","working",12.58,"medium",47.5]],"col_labels":["breed","type","longevity","size","weight"],"row_labels":[4,7,10,17,3,5,6,15]}},"expr":{"type":"SortValuesCall","code":"(dogs[dogs['size'] == 'medium']\n .sort_values('type')\n .groupby('type').median()\n)","range":{"start":{"line":1,"ch":1},"end":{"line":1,"ch":21}}},"annotations":[{"type":"highlight","target":{"table":"lhs","row":"all","col":1}},{"type":"highlight","target":{"table":"rhs","row":"all","col":1}},{"type":"arrow","from":{"table":"lhs","row":0,"col":"all"},"to":{"table":"rhs","row":4,"col":"all"}},{"type":"arrow","from":{"table":"lhs","row":1,"col":"all"},"to":{"table":"rhs","row":0,"col":"all"}},{"type":"arrow","from":{"table":"lhs","row":2,"col":"all"},"to":{"table":"rhs","row":5,"col":"all"}},{"type":"arrow","from":{"table":"lhs","row":3,"col":"all"},"to":{"table":"rhs","row":6,"col":"all"}},{"type":"arrow","from":{"table":"lhs","row":4,"col":"all"},"to":{"table":"rhs","row":1,"col":"all"}},{"type":"arrow","from":{"table":"lhs","row":5,"col":"all"},"to":{"table":"rhs","row":2,"col":"all"}},{"type":"arrow","from":{"table":"lhs","row":6,"col":"all"},"to":{"table":"rhs","row":7,"col":"all"}},{"type":"arrow","from":{"table":"lhs","row":7,"col":"all"},"to":{"table":"rhs","row":3,"col":"all"}}]},{"tables":{"lhs":"prev_rhs","rhs":{"data":[["Bulldog","non-sporting",6.29,"medium",45],["Poodle","non-sporting",11.95,"medium",null],["Boston Terrier","non-sporting",10.92,"medium",null],["French Bulldog","non-sporting",9,"medium",27],["Golden Retriever","sporting",12.04,"medium",60],["Labrador Retriever","sporting",12.04,"medium",67.5],["Boxer","working",8.81,"medium",null],["Siberian Husky","working",12.58,"medium",47.5]],"group_data":{"groups":[{"name":["non-sporting"],"labels":[4,7,10,17]},{"name":["sporting"],"labels":[3,5]},{"name":["working"],"labels":[6,15]}],"col_names":["type"]},"col_labels":["breed","type","longevity","size","weight"],"row_labels":[4,7,10,17,3,5,6,15],"row_groups":["[[\"type\",\"non-sporting\"]]","[[\"type\",\"non-sporting\"]]","[[\"type\",\"non-sporting\"]]","[[\"type\",\"non-sporting\"]]","[[\"type\",\"sporting\"]]","[[\"type\",\"sporting\"]]","[[\"type\",\"working\"]]","[[\"type\",\"working\"]]"],"col_groupkeys":["type"]}},"expr":{"type":"GroupByCall","code":"(dogs[dogs['size'] == 'medium']\n .sort_values('type')\n .groupby('type').median()\n)","range":{"start":{"line":2,"ch":1},"end":{"line":2,"ch":17}}},"annotations":[{"type":"highlight","target":{"table":"lhs","row":"all","col":1}}]},{"tables":{"lhs":"prev_rhs","rhs":{"data":[[9.96,36],[12.04,63.75],[10.695,47.5]],"col_labels":["longevity","weight"],"row_labels":["non-sporting","sporting","working"],"row_label_level_names":["type"]}},"expr":{"type":"AggCall","code":"(dogs[dogs['size'] == 'medium']\n .sort_values('type')\n .groupby('type').median()\n)","range":{"start":{"line":2,"ch":17},"end":{"line":2,"ch":26}}},"annotations":[{"type":"arrow","from":{"table":"lhs","row":0,"col":"all"},"to":{"table":"rhs","row":0,"col":"all"}},{"type":"arrow","from":{"table":"lhs","row":1,"col":"all"},"to":{"table":"rhs","row":0,"col":"all"}},{"type":"arrow","from":{"table":"lhs","row":2,"col":"all"},"to":{"table":"rhs","row":0,"col":"all"}},{"type":"arrow","from":{"table":"lhs","row":3,"col":"all"},"to":{"table":"rhs","row":0,"col":"all"}},{"type":"arrow","from":{"table":"lhs","row":4,"col":"all"},"to":{"table":"rhs","row":1,"col":"all"}},{"type":"arrow","from":{"table":"lhs","row":5,"col":"all"},"to":{"table":"rhs","row":1,"col":"all"}},{"type":"arrow","from":{"table":"lhs","row":6,"col":"all"},"to":{"table":"rhs","row":2,"col":"all"}},{"type":"arrow","from":{"table":"lhs","row":7,"col":"all"},"to":{"table":"rhs","row":2,"col":"all"}}]}],"lang":"py"}`

**Comment by rth** [comment]
Maybe would @pgbovine would be able to answer this question.

**Comment by JothikannanC** [comment]
any answer for this @pgbovine and @rth ?

**Comment by pgbovine** [comment]
Sorry we have not yet made the pandastutor code open-source or packaged up the compiled files yet. You can run it from https://pandastutor.com/ and likely inspect the network traffic to see what files are being loaded. But we currently do not have the development bandwidth to support helping third parties run their own local versions.

**Reply by JothikannanC** [reply]
Oh, I see. The blog I was looking at was saying that the JS library is ready and it's used to replace the backend model. So I thought I might be doing something wrong and asked the question here. Thank you.
